File: engine/src/com/magma/engine/gfx/AnimatedSprite.java
package com.magma.engine.gfx;

import java.util.HashMap;

import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.Animation.PlayMode;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.utils.Array;
import com.magma.engine.assets.MagmaLoader;
import com.magma.engine.utils.Time;

public class AnimatedSprite extends SpriteActor {

	private HashMap<Integer,Animation<AtlasRegion>> animations;
	private int curAnim;
	private boolean isStopped;
	
	public AnimatedSprite(AnimSlot... slots) {
		super(new Sprite(MagmaLoader.getPlaceholderTexture()));
		animations = new HashMap<Integer,Animation<AtlasRegion>>();
		for (AnimSlot slot: slots){
			Array<AtlasRegion> regions = MagmaLoader.getAtlasRegions(slot.name);
			Animation<AtlasRegion> anim = new Animation<AtlasRegion>(1f/slot.fps, regions);
			animations.put(slot.index,anim);
		}
	}

	public void setCurrentAnimation(AnimSlot slot){
		if (!animations.containsKey(slot.index)){
			throw new IllegalArgumentException("Cannot get animation of index " + slot.index + " ("+slot.name+") Not registered!");
		}
		isStopped = false;
		curAnim = slot.index;
	}
	
	public int getCurrentAnimation(){
		return curAnim;
	}

	// TODO: Add more control
	public AnimatedSprite setPlayMode(PlayMode mode) {
		for (Animation<AtlasRegion> anim : animations.values()){
			anim.setPlayMode(mode);
		}
		return this;
	}
	public AnimatedSprite setPlayMode(PlayMode mode, AnimSlot slot) {
		if (!animations.containsKey(slot.index)){
			throw new IllegalArgumentException("Cannot get animation of index " + slot.index + " ("+slot.name+")");
		}
		animations.get(slot.index).setPlayMode(mode);
		return this;
	}

	public PlayMode getPlayMode(AnimSlot slot) {
		if (!animations.containsKey(slot.index)){
			throw new IllegalArgumentException("Cannot get animation of index " + slot.index + " ("+slot.name+")");
		}
		return animations.get(slot.index).getPlayMode();
	}

	public void stop() {
		isStopped = true;
	}
	
	@Override
	public void act(float delta) {
		float time = 0;
		if (!isStopped) {
			time = Time.getTime();
		}
		
		Animation<AtlasRegion> animation = animations.get(curAnim);
		TextureRegion region = null;
		if (animation == null) {
			throw new IllegalStateException("No animation set for " + getClass().getSimpleName());
		}
		
		region = animation.getKeyFrame(time);
		
		// setRegion removes sprites flip, so we need to do this terribleness
		boolean origFlipX = isFlipX();
		boolean origFlipY = isFlipY();
		setRegion(region);
		setFlip(origFlipX, origFlipY);
		super.act(delta);
	}
}

File: engine/src/com/magma/engine/gfx/AnimatedSprite.kt
package com.magma.engine.gfx

import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.Animation.PlayMode
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion
import com.magma.engine.assets.MagmaLoader.getAtlasRegions
import com.magma.engine.assets.MagmaLoader.placeholderTexture
import com.magma.engine.utils.Time

open class AnimatedSprite(vararg slots: AnimSlot) : SpriteActor(Sprite(placeholderTexture)) {
    private val animations: HashMap<Int, Animation<AtlasRegion>> = HashMap()
    private var currentAnimation = 0
    private var isStopped = false

    init {
        for (slot in slots) {
            val regions = getAtlasRegions(slot.name)
            val anim = Animation(1f / slot.fps, regions)
            animations[slot.index] = anim
        }
    }

    fun setCurrentAnimation(slot: AnimSlot) {
        require(animations.containsKey(slot.index)) { "Cannot get animation of index " + slot.index + " (" + slot.name + ") Not registered!" }
        isStopped = false
        currentAnimation = slot.index
    }

    // TODO: Add more control
    fun setPlayMode(mode: PlayMode): AnimatedSprite {
        for (anim in animations.values) {
            anim.playMode = mode
        }
        return this
    }

    fun setPlayMode(mode: PlayMode, slot: AnimSlot): AnimatedSprite {
        require(animations.containsKey(slot.index)) { "Cannot get animation of index ${slot.index} (${slot.name})" }
        animations[slot.index]!!.playMode = mode
        return this
    }

    fun getPlayMode(slot: AnimSlot): PlayMode {
        require(animations.containsKey(slot.index)) { "Cannot get animation of index ${slot.index} (${slot.name})" }
        return animations[slot.index]!!.playMode
    }

    fun stop() {
        isStopped = true
    }

    override fun act(delta: Float) {
        var time = 0f
        if (!isStopped) {
            time = Time.time
        }
        val animation = animations[currentAnimation]
        val region = animation?.getKeyFrame(time)
        checkNotNull(animation) { "No animation set for ${javaClass.simpleName}" }

        // setRegion removes sprites flip, so we need to do this terribleness
        val origFlipX = isFlipX
        val origFlipY = isFlipY
        setSpriteRegion(region!!)
        setFlip(origFlipX, origFlipY)

        super.act(delta)
    }
}